File: heaps/shortest_paths/shortest_paths.cpp
#include <vector>
#include <queue>
using namespace std;

class ShortestPathsCalculator
{ // LeetCode Q.2642.
private:
    vector<vector<int>> shortest_paths;
    int total_nodes;
    int infinity = numeric_limits<int>::max();

public:
    ShortestPathsCalculator(int n, vector<vector<int>> &edges)
    {
        shortest_paths.assign(n, vector<int>(n, infinity));
        for (auto edge : edges)
        { // Edge format: {out node, in node, cost}.
            shortest_paths[edge[0]][edge[1]] = edge[2];
        }
        for (int node = 0; node < n; node++)
        {
            shortest_paths[node][node] = 0; // Each node's cost to itself is 0.
        }
        total_nodes = n;
    }

    void add_edge(vector<int> edge)
    { // Edge format: {out node, in node, cost}.
        shortest_paths[edge[0]][edge[1]] = edge[2];
    }

    int calculate_shortest_path(int start, int end)
    {
        // Min heap. Format: {cost, node}.
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> cost_heap;

        for (int node = 0; node < total_nodes; node++)
        {
            if (start != node && shortest_paths[start][node] != infinity)
            {
                cost_heap.push({shortest_paths[start][node], node});
            }
        }

        while (!cost_heap.empty() && cost_heap.top().first != infinity)
        {
            auto [min_cost, min_cost_node] = cost_heap.top();
            cost_heap.pop();

            for (int neighbor = 0; neighbor < total_nodes; neighbor++)
            {
                int cost = shortest_paths[min_cost_node][neighbor];
                if (cost != infinity)
                {
                    if (min_cost + cost < shortest_paths[start][neighbor])
                    {
                        shortest_paths[start][neighbor] = min_cost + cost;
                        cost_heap.push({shortest_paths[start][neighbor], neighbor});
                    }
                }
            }
        }

        if (shortest_paths[start][end] == infinity)
        {
            return -1;
        }
        return shortest_paths[start][end];
    }
};

File: heaps/shortest_paths/shortest_paths.py
import heapq


class ShortestPathsCalculator:  # LeetCode Q.2642.
    def __init__(self, total_nodes: int, edges: list[list[int]]):
        self.infinity = float("inf")
        self.shortest_paths: list[list[int | float]] = [
            [self.infinity] * total_nodes for _ in range(total_nodes)
        ]
        for out_node, in_node, cost in edges:  # Directed edges.
            self.shortest_paths[out_node][in_node] = cost

        for node in range(total_nodes):  # Each node's cost to itself is 0.
            self.shortest_paths[node][node] = 0

    def add_edge(self, edge: list[int]) -> None:
        out_node, in_node, cost = edge
        self.shortest_paths[out_node][in_node] = cost

    def calculate_shortest_path(self, start: int, end: int) -> int:
        cost_heap = []  # Min heap. Format: (cost, node).
        for in_node, cost in enumerate(self.shortest_paths[start]):
            if in_node != start and cost != self.infinity:
                heapq.heappush(cost_heap, (cost, in_node))

        while cost_heap and cost_heap[0][0] != self.infinity:
            min_cost, min_cost_node = heapq.heappop(cost_heap)

            for neighbor, cost in enumerate(self.shortest_paths[min_cost_node]):
                if min_cost + cost < self.shortest_paths[start][neighbor]:
                    self.shortest_paths[start][neighbor] = min_cost + cost
                    heapq.heappush(
                        cost_heap, (self.shortest_paths[start][neighbor], neighbor)
                    )

        if self.shortest_paths[start][end] == self.infinity:
            return -1
        return self.shortest_paths[start][end]
